src/UI/PlayerSetupController.java

```diff
@@ -0,0 +1,208 @@
+  private OptionSelector playerSelector;
+  private OptionSelector categorySelector;
+
+  private GameBuilder gameBuilder = null;
+  PlayerSetupInfo[] coSelectors;
+
+  public enum SelectionCategories { COMMANDER, COLOR_FACTION, TEAM, AI, START };
+
+  private IController subMenu;
+
+  public PlayerSetupController( GameBuilder builder )
+  {
+    // Once we hit go, we plug all the COs we chose into our gameBuilder.
+    gameBuilder = builder;
+
+    // Set up our row/col selectors.
+    int numCos = gameBuilder.mapInfo.getNumCos();
+    playerSelector = new OptionSelector(numCos);
+    categorySelector = new OptionSelector(SelectionCategories.values().length);
+    categorySelector.setSelectedOption(SelectionCategories.START.ordinal()); // Best case is that no changes are needed.
+
+    // Create objects to keep track of the selected options for each player.
+    coSelectors = new PlayerSetupInfo[numCos];
+
+    // Start by making default CO/color selections.
+    for(int co = 0; co < numCos; ++co)
+    {
+      // Set up our option selection framework
+      coSelectors[co] = new PlayerSetupInfo(numCos, co, CommanderLibrary.getCommanderList(), UIUtils.getCOColors(), UIUtils.getFactions(), AILibrary.getAIList());
+    }
+  }
+
+  @Override
+  public boolean handleInput(InputAction action)
+  {
+    boolean exitMenu = false;
+    if(null != subMenu)
+    {
+      boolean exitSub = subMenu.handleInput(action);
+      if(exitSub)
+      {
+        // Only apply the change from the sub-menu if it exited via ENTER.
+        if(action == InputAction.ENTER)
+        {
+          final int category = categorySelector.getSelectionNormalized();
+          if( category == SelectionCategories.COMMANDER.ordinal() )
+          {
+            PlayerSetupCommanderController cmdrMenu = (PlayerSetupCommanderController)subMenu;
+            PlayerSetupInfo info = getPlayerInfo(playerSelector.getSelectionNormalized());
+            info.currentCO.setSelectedOption(cmdrMenu.getSelectedCommander());
+          }
+          if( category == SelectionCategories.COLOR_FACTION.ordinal() )
+          {
+            PlayerSetupColorFactionController unitMenu = (PlayerSetupColorFactionController)subMenu;
+            PlayerSetupInfo info = getPlayerInfo(playerSelector.getSelectionNormalized());
+            info.currentColor.setSelectedOption(unitMenu.getSelectedColor());
+            info.currentFaction.setSelectedOption(unitMenu.getSelectedFaction());
+          }
+          if( category == SelectionCategories.TEAM.ordinal() )
+          { /* Do nothing - the sub-menu does the change on its own. */ }
+          if( category == SelectionCategories.AI.ordinal() )
+          {
+            PlayerSetupAiController aiMenu = (PlayerSetupAiController)subMenu;
+            PlayerSetupInfo info = getPlayerInfo(playerSelector.getSelectionNormalized());
+            info.currentAI.setSelectedOption(aiMenu.getSelectedAiIndex());
+          }
+        }
+        subMenu = null;
```

Comment:
Hmm. This could totally be a switch/case, but I think I like the idea of deleting it all, heh.

Reply:
Deleted it all.